include/pmacc/lockstep/WorkerCfg.hpp

```diff
@@ -0,0 +1,117 @@
+/* Copyright 2022 Rene Widera
+ *
+ * This file is part of PMacc.
+ *
+ * PMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc/lockstep/Worker.hpp"
+#include "pmacc/traits/GetNumWorkers.hpp"
+#include "pmacc/types.hpp"
+
+#include "pmacc/math/vector/compile-time/Vector.hpp"
+
+
+namespace pmacc::lockstep
+{
+    namespace detail
+    {
+        struct WorkerCfgAssume1DThreads
+        {
+            template<typename T_WorkerCfg, typename T_Acc>
+            HDINLINE static auto getWorker(T_Acc const& acc)
+            {
+                return T_WorkerCfg::getWorkerAssume1DThreads(acc);
+            }
+        };
+    } // namespace detail
+    /** Configuration of worker used for a lockstep kernel
+     *
+     * @tparam T_numWorkers Number of workers which are used to execute this functor.
+     *
+     * @attention: The real number of workers used for the lockstep kernel depends on the alpaka backend and will
+     * be adjusted by this class via the trait pmacc::traits::GetNumWorkers.
+     */
+    template<uint32_t T_numWorkers>
+    struct WorkerCfg
+    {
+        friend struct detail::WorkerCfgAssume1DThreads;
+
+        //! number of workers
+        static constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<T_numWorkers>::value;
+
+        /** get the worker index
+         *
+         * @return index of the worker
+         */
+        template<typename T_Acc>
+        HDINLINE static auto getWorker(T_Acc const& acc)
+        {
+            auto const localThreadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc);
+            auto const blockExtent = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc);
+            auto const linearThreadIdx = alpaka::mapIdx<1u>(localThreadIdx, blockExtent)[0];
+            return Worker<T_Acc, T_numWorkers>(acc, linearThreadIdx);
```

Comment:
There is no check, including in `Worker`, whether `T_numWorkers` actually matches `blockExtent.prod()`. Routine runtime check here would be feasible, but there should at least be an assertion for debug builds somewhere, maybe an `ALPAKA_ASSERT_OFFLOAD(...)`.

Reply:
Yes I also thought about it. I will use your suggestion. The check should only be enabled if we compile in debug mode else the performance will be bad and `assert` on GPU is requiring registers what is bad too. For a debug compile this is fine.

Reply:
I added the check.